sql/drop.go

```diff
@@ -22,10 +22,147 @@ import (
 
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/proto"
 	"github.com/cockroachdb/cockroach/sql/parser"
 	"github.com/cockroachdb/cockroach/sql/privilege"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
 )
 
+// DropDatabase drops a database.
+// Privileges: DROP on database.
+//   Notes: postgres allows only the database owner to DROP a database.
+//          mysql requires the DROP privileges on the database.
+// TODO(XisiHuang): our DROP DATABASE is like the postgres DROP SCHEMA
+// (cockroach database == postgres schema). the postgres default of not
+// dropping the schema if there are dependent objects is more sensible
+// (see the RESTRICT and CASCADE options).
+func (p *planner) DropDatabase(n *parser.DropDatabase) (planNode, error) {
+	if n.Name == "" {
+		return nil, errEmptyDatabaseName
+	}
+
+	nameKey := MakeNameMetadataKey(keys.RootNamespaceID, string(n.Name))
+	gr, err := p.txn.Get(nameKey)
+	if err != nil {
+		return nil, err
+	}
+	if !gr.Exists() {
+		if n.IfExists {
+			// Noop.
+			return &valuesNode{}, nil
+		}
+		return nil, fmt.Errorf("database %q does not exist", n.Name)
+	}
+
+	descKey := MakeDescMetadataKey(ID(gr.ValueInt()))
+	desc := DatabaseDescriptor{}
+	if err := p.txn.GetProto(descKey, &desc); err != nil {
+		return nil, err
+	}
+	if err := desc.Validate(); err != nil {
+		return nil, err
+	}
+
+	if err := p.checkPrivilege(&desc, privilege.DROP); err != nil {
+		return nil, err
+	}
+
+	tbNames, err := p.getTableNames(&desc)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := p.DropTable(&parser.DropTable{Names: tbNames}); err != nil {
+		return nil, err
+	}
+
+	b := &client.Batch{}
+	b.Del(descKey)
+	b.Del(nameKey)
+	// Delete the zone config entry for this database.
+	b.Del(MakeZoneKey(desc.ID))
+	// Mark transaction as operating on the system DB.
+	p.txn.SetSystemDBTrigger()
+	if err := p.txn.Run(b); err != nil {
+		return nil, err
+	}
+	return &valuesNode{}, nil
+}
+
+// DropIndex drops an index.
+// Privileges: CREATE on table.
+//   Notes: postgres allows only the index owner to DROP an index.
+//          mysql requires the INDEX privilege on the table.
+func (p *planner) DropIndex(n *parser.DropIndex) (planNode, error) {
+	var b client.Batch
+
+	for _, indexQualifiedName := range n.Names {
+		if err := indexQualifiedName.NormalizeTableName(p.session.Database); err != nil {
+			return nil, err
+		}
+
+		tableDesc, err := p.getTableDesc(indexQualifiedName)
+		if err != nil {
+			return nil, err
+		}
+
+		idxName := indexQualifiedName.Index()
+		idx, err := tableDesc.FindIndexByName(idxName)
+		if err != nil {
+			if n.IfExists {
+				// Noop.
+				return &valuesNode{}, nil
+			}
+			// index does not exist, but we want it to: error out.
+			return nil, fmt.Errorf("index %q does not exist", idxName)
```

Comment:
What other comment? The one about the scoping of index names? I'm pretty comfortable stating that index names should be scoped to tables until we hear a strong argument otherwise (and simply following postgres behavior isn't a strong argument for me).